src/py21cmfast/drivers/coeval.py

```diff
@@ -605,44 +598,41 @@ def run_coeval(
     singleton = False
     # Ensure perturb is a list of boxes, not just one.
     if perturbed_field is None:
-        perturbed_field = []
+        perturbed_field = ()
     elif not hasattr(perturbed_field, "__len__"):
-        perturbed_field = [perturbed_field]
+        perturbed_field = (perturbed_field,)
         singleton = True
 
-    random_seed = (
-        initial_conditions.random_seed
-        if initial_conditions is not None
-        else random_seed
+    # ensure inputs are compatible with ICs/Perturbedfields
+    inputs = InputParameters.from_output_structs(
+        (initial_conditions,) + perturbed_field,
+        cosmo_params=inputs.cosmo_params,
+        user_params=inputs.user_params,
+        astro_params=inputs.astro_params,
+        flag_options=inputs.flag_options,
+        random_seed=inputs.random_seed,
```

Comment:
I reckon we should just write a new method on `InputParameters` that does the compatibility check (even if it itself just wraps the `.from_output_structs`). Something with a better name, like `inputs.check_compatibility(outputs)`.